modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkIndexByScrollAction.java

```diff
@@ -40,216 +41,359 @@
+                                                ActionListener<BulkIndexByScrollResponse> listener) {
         super(task, logger, client, threadPool, mainRequest, firstSearchRequest, listener);
-        this.scriptService = scriptService;
-        if (mainRequest.getScript() == null) {
-            script = null;
-        } else {
-            script = scriptService.compile(mainRequest.getScript(), ScriptContext.Standard.UPDATE, emptyMap(), state);
-        }
     }
 
     @Override
-    protected BulkIndexByScrollResponse buildResponse(TimeValue took, List<BulkItemResponse.Failure> indexingFailures,
-                                     List<ShardSearchFailure> searchFailures, boolean timedOut) {
-        return new BulkIndexByScrollResponse(took, task.getStatus(), indexingFailures, searchFailures, timedOut);
-    }
-
-    /**
-     * Build the IndexRequest for a single search hit. This shouldn't handle
-     * metadata or the script. That will be handled by copyMetadata and
-     * applyScript functions that can be overridden.
-     */
-    protected abstract IndexRequest buildIndexRequest(SearchHit doc);
-
-    @Override
     protected BulkRequest buildBulk(Iterable<SearchHit> docs) {
         BulkRequest bulkRequest = new BulkRequest();
-        ExecutableScript executableScript = null;
-        Map<String, Object> scriptCtx = null;
-
         for (SearchHit doc : docs) {
-            if (doc.hasSource()) {
-                /*
-                 * Either the document didn't store _source or we didn't fetch it for some reason. Since we don't allow the user to
-                 * change the "fields" part of the search request it is unlikely that we got here because we didn't fetch _source.
-                 * Thus the error message assumes that it wasn't stored.
-                 */
-                throw new IllegalArgumentException("[" + doc.index() + "][" + doc.type() + "][" + doc.id() + "] didn't store _source");
-            }
-            IndexRequest index = buildIndexRequest(doc);
-            copyMetadata(index, doc);
-            if (script != null) {
-                if (executableScript == null) {
-                    executableScript = scriptService.executable(script, mainRequest.getScript().getParams());
-                    scriptCtx = new HashMap<>();
-                }
-                if (false == applyScript(index, doc, executableScript, scriptCtx)) {
-                    continue;
+            if (accept(doc)) {
+                ActionRequest<?> request = apply(buildRequest(doc), doc);
+                if (request != null) {
+                    bulkRequest.add(request);
                 }
             }
-            bulkRequest.add(index);
         }
-
         return bulkRequest;
     }
 
     /**
-     * Copies the metadata from a hit to the index request.
+     * Used to accept or ignore a search hit. Ignored search hits will be excluded
+     * from the bulk request.
```

Comment:
Might want to mention that this is also where we fail on invalid search hits. I expect delete-by-query doesn't need to fail like this though.